src/webgpu/api/validation/resource_usages/textureUsageInRender.spec.ts

```diff
@@ -574,3 +577,68 @@ g.test('bindings_in_bundle')
       encoder.finish();
     }, !success);
   });
+
+g.test('unused_bindings_in_pipeline')
+  .params(params().combine(pbool('useBindGroup0')).combine(pbool('useBindGroup1')))
+  .fn(async t => {
+    const { useBindGroup0, useBindGroup1 } = t.params;
+    const view = t.createTexture({ usage: GPUTextureUsage.STORAGE }).createView();
+    const bindGroup0 = t.createBindGroup(0, view, 'readonly-storage-texture', 'rgba8unorm');
+    const bindGroup1 = t.createBindGroup(0, view, 'writeonly-storage-texture', 'rgba8unorm');
+
+    const wgslVertex = `
+      fn main() -> void {
+      return;
+      }
+
+      entry_point vertex = main;
+    `;
+    const binding0: string = useBindGroup0 ? '[[set 0, binding 0]] var<image> image0' : '';
+    const binding1: string = useBindGroup1 ? '[[set 1, binding 0]] var<image> image1' : '';
+    const wgslFragment =
+      binding0 +
+      binding1 +
+      `
+      fn main() -> void {
+        return;
+      }
+
+      entry_point fragment = main;
+    `;
+    const pipeline = t.device.createRenderPipeline({
+      vertexStage: {
+        module: t.device.createShaderModule({
+          code: wgslVertex,
+        }),
+        entryPoint: 'main',
+      },
+      fragmentStage: {
+        module: t.device.createShaderModule({
+          code: wgslFragment,
+        }),
+        entryPoint: 'main',
+      },
+      primitiveTopology: 'triangle-list',
+      colorStates: [{ format: 'rgba8unorm' }],
+    });
+
+    const encoder = t.device.createCommandEncoder();
+    const pass = encoder.beginRenderPass({
+      colorAttachments: [
+        {
+          attachment: t.createTexture().createView(),
+          loadValue: { r: 0.0, g: 1.0, b: 0.0, a: 1.0 },
+          storeOp: 'store',
+        },
+      ],
+    });
+    pass.setPipeline(pipeline);
+    pass.setBindGroup(0, bindGroup0);
+    pass.setBindGroup(1, bindGroup1);
+    pass.draw(3, 1, 0, 0);
```

Comment:
Technically this test should fail even if neither `setPipeline` nor `draw` is set at all. Since we're trying to tickle errors in validation how about testing a few additional permutations here, like:
- setPipeline, setBindGroups, draw
- setPipeline, setBindGroups
- setBindGroups, setPipeline
- setPipeline, setBindGroups, setPipeline

x= setBindGroups {[0, 1], [1, 0]}

params could be:
- setPipelineBefore = {true,false}
- setBindGroupsOrder = {[0, 1], [1, 0]}
- setPipelineAfter = {true, false}
- draw = {true, false}

Reply:
Per my understanding, setting Pipeline before setting bind groups or after setting bind groups should be tested in pipeline (or bind group) related tests. And I remember we do have those tests in Dawn validation tests in pipeline (or bindgroups?) related tests, which manifest that the sequence doesn't change anything. So I suggest we ignore the sequence here. But adding setPipeline and draw as bool parameters to the tests is a really good suggestion, in addition to adding bindings or not in shader code. Right now, the tests can manifest that we should validate resources if they are set via state setting API (like setBindGroup) regardless of whether the shader, or pipeline, or draw call depends on it.

Reply:
It can't hurt to have it here too, since like @kainino0x said the goal is to tickle the resource tracking validation and we're combining with other interesting tickly combinations.

Reply:
Yes, and these are fast tests so there's no real harm in there being possible duplication. Especially since it's very easy to express these tests here. You can even add cases like setBindGroupsOrder [0, 1, 0].

Reply:
No problem. I have done this change. I am not sure how to do  "setBindGroupsOrder [0, 1, 0]" or "setBindGroupsOrder = {[0, 1], [1, 0]}". But I suppose that you want to reverse the bind groups indices. In addition, setPipeline can be added into the middle of two setBindGroups, which make sense for the index order of bind groups too.

Reply:
With the addition you made to setPipeline between the two calls, you're right setBindGroupsOrder no longer makes sense. Your latest change LGTM.